Update jiosaavn.py

alien-developers opened the pull request:
updated code to pull highest quality 320 kbps

**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

ADD DESCRIPTION HERE

Fixes #


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

seproDev commented:
Please fill out the PR template properly, otherwise this can not be merged.

There are five quality levels available: 16, 32, 64, 128, 320
Does it make sense to request all five and list them as formats?

alien-developers commented:
> Please fill out the PR template properly, otherwise this can not be merged.
> 
> There are five quality levels available: 16, 32, 64, 128, 320 Does it make sense to request all five and list them as formats?

You mean Instead of selecting highquality, its better to list formats and select one of the,. Am icorrect?

seproDev commented:
Yes that would be the idea. It would come at the cost of making 5 instead of 1 HTTP request. I think that's fine, but maybe others disagree. I would at least request both the 128 and 320 format.

alien-developers commented:
Sure i will work on that accordingly

bashonly commented:
it might be appropriate to have this configurable by extractor-arg, e.g:
```py
        formats = []
        # available bitrates are 16, 32, 64, 128, 320
        for bitrate in self._configuration_arg('bitrate', ['128', '320'], ie_key='JioSaavn'):
            media_data = self._download_json(
                'https://www.jiosaavn.com/api.php', audio_id, f'Downloading format info for {bitrate}',
                fatal=False, data=urlencode_postdata({
                    '__call': 'song.generateAuthToken',
                    '_format': 'json',
                    'bitrate': bitrate,
                    'url': song_data['encrypted_media_url'],
                }))
            if not media_data.get('auth_url'):
                self.report_warning(f'Unable to extract format info for {bitrate}')
                continue
            formats.append({
                'url': media_data['auth_url'],
                'ext': media_data.get('type'),
                'format_id': bitrate,
                'abr': int(bitrate),
                'vcodec': 'none',
            })
```

alien-developers commented:
Hey thanks for your suggested. I have updated my code accordingly. instead on two i have added all formats and we can use to below command to download certain format song.

--match-filter "abr = 320"

alien-developers commented:
Below is the output i got when i use '-f 320' for downloading songs in 320kbps'

[log.txt](https://github.com/yt-dlp/yt-dlp/files/13764458/log.txt)
![Screenshot 2023-12-25 121420](https://github.com/yt-dlp/yt-dlp/assets/154035958/a1fdfbb0-d58a-436b-b698-6d5f93cc3b65)

seproDev commented on yt_dlp/extractor/jiosaavn.py:
```diff
@@ -36,29 +36,39 @@ class JioSaavnSongIE(JioSaavnBaseIE):
+            if not media_data.get('auth_url'):
+                self.report_warning(f'Unable to extract format info for {bitrate}')
+                continue
+            formats.append({
+                'url': media_data['auth_url'],
+                'ext': media_data.get('type'),
+                'id': bitrate,
+                'format_id': bitrate,
+                'abr': int(bitrate),
+                'vcodec': 'none',
+                'acodec': media_data.get('type'),
+                **traverse_obj(song_data, {
+                    'title': ('title', 'text'),
+                    'album': ('album', 'text'),
+                    'thumbnail': ('image', 0, {url_or_none}),
+                }),
+            })
+        return self.playlist_result(formats)
```
These are formats of a single video and not different videos of a playlist.
We can also add more metadata.
```suggestion
            formats.append({
                'url': media_data['auth_url'],
                'ext': media_data.get('type'),
                'format_id': bitrate,
                'abr': int(bitrate),
                'vcodec': 'none',
            })
        return {
            'id': audio_id,
            'formats': formats,
            **traverse_obj(song_data, {
                'title': ('title', 'text'),
                'album': ('album', 'text'),
                'thumbnail': ('image', 0, {url_or_none}),
                'duration': ('duration', {int_or_none}),
                'view_count': ('play_count', {int_or_none}),
                'release_year': ('year', {int_or_none}),
            }),
        }
```
Tests will need to be updated and `int_or_none` imported from `..utils`.

seproDev commented on yt_dlp/extractor/jiosaavn.py:
```diff
@@ -36,29 +36,39 @@ class JioSaavnSongIE(JioSaavnBaseIE):
-                'thumbnail': ('image', 0, {url_or_none}),
-            }),
-        }
+        formats = []
+        # available bitrates are 16, 32, 64, 128, 320
+        for bitrate in self._configuration_arg('bitrate', ['16', '32', '64', '128', '320'], ie_key='JioSaavn'):
```
The idea here was that we don't request all bitrates in order to speed up downloading, and instead allow the user to extract additional formats if they so desire. I would also add a check to see if the provided bitrate was valid (the server returns a file, regardless of what you request) and change the order:
```suggestion
        extract_bitrates = self._configuration_arg('bitrate', ['128', '320'], ie_key='JioSaavn')
        if not all(bitrate in ('16', '32', '64', '128', '320') for bitrate in extract_bitrates):
            raise ValueError(f'Invalid bitrate(s) {extract_bitrates}')

        song_data = self._extract_initial_data(url, audio_id)['song']['song']
        formats = []
        for bitrate in extract_bitrates:
```
Please also document this extractor arg in the readme.

seproDev commented on yt_dlp/extractor/jiosaavn.py:
```diff
@@ -36,29 +36,39 @@ class JioSaavnSongIE(JioSaavnBaseIE):
     def _real_extract(self, url):
         audio_id = self._match_id(url)
         song_data = self._extract_initial_data(url, audio_id)['song']['song']
```
With the changes from above, this line should be removed.
```suggestion
```

seproDev commented on yt_dlp/extractor/jiosaavn.py:
```diff
@@ -36,29 +36,39 @@ class JioSaavnSongIE(JioSaavnBaseIE):
 
 class JioSaavnAlbumIE(JioSaavnBaseIE):
-    _VALID_URL = r'https?://(?:www\.)?(?:jio)?saavn\.com/album/[^/?#]+/(?P<id>[^/?#]+)'
+    _VALID_URL = r'https?://(?:www\.)?(?:jio)?saavn\.com/(?:s/playlist|album)/[^/?#]+/(?P<id>[^/?#]+)'
```
Are you sure this is correct? From what I can see, playlists work differently than albums.
A test would also need to be added.

bashonly commented on yt_dlp/extractor/jiosaavn.py:
```diff
@@ -35,24 +39,42 @@ class JioSaavnSongIE(JioSaavnBaseIE):
 
     def _real_extract(self, url):
```
```suggestion

    _VALID_BITRATES = ('16', '32', '64', '128', '320')

    def _real_extract(self, url):
```

bashonly commented on yt_dlp/extractor/jiosaavn.py:
```diff
@@ -35,24 +39,42 @@ class JioSaavnSongIE(JioSaavnBaseIE):
-                'bitrate': '128',
-                'url': song_data['encrypted_media_url'],
-            }))
 
+        extract_bitrates = self._configuration_arg('bitrate', ['128', '320'], ie_key='JioSaavn')
+        if not all(bitrate in ('16', '32', '64', '128', '320') for bitrate in extract_bitrates):
+            raise ValueError(f'Invalid bitrate(s) {extract_bitrates}')
```
```suggestion
        extract_bitrates = self._configuration_arg('bitrate', ['128', '320'], ie_key='JioSaavn')
        if invalid_bitrates := [br for br in extract_bitrates if br not in self._VALID_BITRATES]:
            raise ValueError(
                f'Invalid bitrate(s): {", ".join(invalid_bitrates)}. '
                + f'Valid bitrates are: {", ".join(self._VALID_BITRATES)}')
```

bashonly commented on README.md:
```diff
@@ -1888,6 +1888,9 @@ The following extractors use this feature:
 * `type`: Type(s) of game replays to extract. Valid types are: `full_game`, `full_game_spanish`, `condensed_game` and `all_22`. You can use `all` to extract all available replay types, which is the default
 
+#### jiosaavn
+* `bitrate`: Which bitrates to request. Valid options are `16`, `32`, `64`, `128`, and `320` or a comma-separated list of these values. Defaults to `128,320`.
```
```suggestion
* `bitrate`: Audio bitrates to request. One or more of `16`, `32`, `64`, `128`, or `320`. Default is `128,320`
```

bashonly commented on yt_dlp/extractor/jiosaavn.py:
```diff
@@ -20,39 +21,63 @@ class JioSaavnSongIE(JioSaavnBaseIE):
+                'format_id': bitrate,
+                'abr': int(bitrate),
+                'vcodec': 'none',
+            })
         return {
```
```suggestion
            })

        return {
```
